commitment_scheme/src/merkle/mod.rs

```diff
@@ -0,0 +1,2014 @@
+pub mod hash;
+pub mod merkle_commitment_scheme;
+use crate::merkle::hash::Hasher;
+use generic_array::GenericArray;
+use thiserror::Error;
+
+use std::{collections::VecDeque, marker::PhantomData};
+
+use ark_ff::PrimeField;
+
+use channel::ProverChannel;
+use channel::VerifierChannel;
+use channel::{fs_prover_channel::FSProverChannel, fs_verifier_channel::FSVerifierChannel};
+use generic_array::typenum::U32;
+use randomness::Prng;
+use sha3::Digest;
+
+pub struct MerkleTree<F: PrimeField, H: Hasher<F>> {
+    data_length: usize,
+    nodes: Vec<H::Output>,
+    _ph: PhantomData<F>,
+}
+
+impl<F: PrimeField, H: Hasher<F>> MerkleTree<F, H> {
+    /// Constructs a new MerkleTree with empty nodes.
+    ///
+    /// # Arguments
+    ///
+    /// - `data_length`: length of data which the tree will commit to.
+    ///
+    /// # Returns
+    ///
+    /// - `Self`: MerkleTree with empty nodes
+    pub fn new(data_length: usize) -> Self {
+        // assert data_length is a power of 2
+        assert!(
+            data_length != 0 && (data_length & (data_length - 1)) == 0,
+            "Data length is not a power of two."
+        );
+
+        // for better indexing, use an array that has one extra cell at the beginning
+        let nodes: Vec<H::Output> = vec![H::Output::default(); 2 * data_length];
+
+        Self {
+            data_length,
+            nodes,
+            _ph: PhantomData,
+        }
+    }
+}
+
+impl<F: PrimeField, H: Hasher<F, Output = GenericArray<u8, U32>>> MerkleTree<F, H> {
+    /// Feeds the tree with data to commit on.
+    ///
+    /// # Arguments
+    ///
+    /// - `data`: input data to commit on.
+    /// - `start_index`: index of the tree leaves where the data will be added.
+    pub fn add_data(&mut self, data: &[H::Output], start_index: usize) {
+        assert!(start_index + data.len() <= self.data_length);
+
+        // copy given data to the leaves of the tree
+        self.nodes.splice(
+            (self.data_length + start_index)..(self.data_length + start_index + data.len()),
+            data.to_vec(),
+        );
+
+        // Hash to compute all parent nodes
+        let mut cur = (self.data_length + start_index) / 2;
+        let mut sub_layer_length = data.len() / 2;
+
+        while sub_layer_length > 0 {
+            for j in cur..(cur + sub_layer_length) {
+                self.nodes[j] = H::node(&[self.nodes[2 * j], self.nodes[2 * j + 1]]);
+            }
+            cur /= 2;
+            sub_layer_length /= 2;
+        }
+    }
+
+    /// Retrieves the root of the tree assuming nodes at some depth are known.
+    ///
+    /// # Arguments
+    ///
+    /// - `min_depth_assumed_correct`: minimal depth (distance from the root) assumed to be correct.
+    ///
+    /// # Returns
+    ///
+    /// Returns the root of the tree.
+    pub fn get_root(&mut self, min_depth_assumed_correct: usize) -> H::Output {
```

Comment:
In the implementation, the `get_root` function is only called with `min_depth_assumed_correct` equal to height of the tree, even though in the test cases they are testing for different values of  `min_depth_assumed_correct`.

Since it is only called with `min_depth_assumed_correct` equal to height, I was thinking to remove the `min_depth_assumed_correct` argument from the function and update the function to be as follows:

```
pub fn get_root(&mut self) -> H::Output {
    let height_correct = self.data_length;
    
    for j in (1..height_correct).rev() {
        self.nodes[j] = H::node(&[self.nodes[2 * j].clone(), self.nodes[2 * j + 1].clone()]);
    }
    
    self.nodes[1].clone()
}
```
The updated function computes the root starting from the leaf nodes, so the above case of computing root from different depths will not occur.